kernel/comps/virtio/src/device/socket/device.rs

```diff
@@ -0,0 +1,342 @@
+        let header = VirtioVsockHdr {
+            op: VirtioVsockOp::Rw as u16,
+            len,
+            ..connection_info.new_header(self.guest_cid)
+        };
+        connection_info.tx_cnt += len;
+        self.send_packet_to_tx_queue(&header, buffer)
+    }
+
+    /// Receive bytes from peer, returns the header
+    pub fn receive(
+        &mut self,
+        buffer: &mut [u8],
+        // connection_info: &mut ConnectionInfo,
+    ) -> Result<VirtioVsockHdr, SocketError> {
+        let (token, len) = self.recv_queue.pop_used()?;
+        debug!(
+            "receive packet in rx_queue: token = {}, len = {}",
+            token, len
+        );
+        let mut rx_buffer = self
+            .rx_buffers
+            .remove(token as usize)
+            .ok_or(QueueError::WrongToken)?;
+        rx_buffer.set_packet_len(RX_BUFFER_SIZE);
+
+        let (header, payload) = read_header_and_body(rx_buffer.buf())?;
+        // The length written should be equal to len(header)+len(packet)
+        assert_eq!(len, header.len() + VIRTIO_VSOCK_HDR_LEN as u32);
+        debug!("Received packet {:?}. Op {:?}", header, header.op());
+        debug!("body is {:?}", payload);
+
+        assert!(buffer.len() >= payload.len());
+        buffer[..payload.len()].copy_from_slice(payload);
+
+        self.add_rx_buffer(rx_buffer, token)?;
+        Ok(header)
+    }
+
+    /// Polls the RX virtqueue for the next event, and calls the given handler function to handle it.
+    pub fn poll(
+        &mut self,
+        handler: impl FnOnce(VsockEvent, &[u8]) -> Result<Option<VsockEvent>, SocketError>,
+    ) -> Result<Option<VsockEvent>, SocketError> {
+        // Return None if there is no pending packet.
+        if !self.recv_queue.can_pop() {
+            return Ok(None);
+        }
+        let mut body = RxBuffer::new(RX_BUFFER_SIZE);
+        let header = self.receive(body.buf_mut())?;
+
+        VsockEvent::from_header(&header).and_then(|event| handler(event, body.buf()))
+    }
+
+    /// Add a used rx buffer to recv queue,@index is only to check the correctness
+    fn add_rx_buffer(&mut self, mut rx_buffer: RxBuffer, index: u16) -> Result<(), SocketError> {
+        let token = self.recv_queue.add_buf(&[], &[rx_buffer.buf_mut()])?;
+        assert_eq!(index, token);
+        assert!(self.rx_buffers.put_at(token as usize, rx_buffer).is_none());
+        if self.recv_queue.should_notify() {
+            self.recv_queue.notify();
+        }
+        Ok(())
+    }
+
+    /// Negotiate features for the device specified bits 0~23
+    pub(crate) fn negotiate_features(features: u64) -> u64 {
+        let device_features = VsockFeatures::from_bits_truncate(features);
+        let supported_features = VsockFeatures::supported_features();
+        let vsock_features = device_features & supported_features;
+        debug!("features negotiated: {:?}", vsock_features);
+        vsock_features.bits()
+    }
+}
+
+impl Debug for SocketDevice {
```

Comment:
Does this Debug implementation have something special? Why not just using `#[derive(Debug)]`?

Reply:
Just because other virtio devices also implement Debug manually.